keps/sig-auth/0000-authority-delegation.md

```diff
@@ -0,0 +1,259 @@
+A new non-namespaced persistent object will be added to the
+`authorization.k8s.io` API group to store instances of authority delegation.
+
+```go
+type AuthoritiyDelegation struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // The user that created the delegation.
+  User Info
+  // The principal that is delegating authority.
+  Subject SubjectRef
+  // The principals that are delegated authority.
+  Delegates []SubjectRef
+  // defined above
+  Authority AuthorityReference
+  // If this delegation is created selecting a previously delegated authority,
+  // the delegation chain is recorded here.
+  DelegationHistory []AuthorityDelegationReference
+}
+
+type SubjectRef struct {
+  Kind SubjectKind
+  Name string
+  UID types.UID
+}
+
+const (
+  GroupKind SubjectKind = "Group"
+  UserKind SubjectKind = "User"
+)
+
+type AuthrityDelegationReference struct {
+  UID types.UID
+}
+```
+
+#### AuthorityDelegationPolicy
+
+A new policy will be introduced to implictly create `AuthorityDelegations` in
+reaction to API actions. The `AuthorityDelegationPolicy` will be a
+non-namespaced object in the `authorization.k8s.io` API group.
+
+```go
+type AuthorityDelegationPolicy struct {
+  // The resource attributes of an action to implicitly create a delegation for.
+  ResourceAttributes *ResourceAttributes
+  // The principals that are delegated authority.
+  Delegates []SubjectRef
+  // The permissions that are delegated.
+  Authority AuthorityReference
+}
+```
+
+When an API action matching an `AuthorityDelegationPolicy` is processed, an
+`AuthorityDelegation` is implicitly created before any objects are persisted.
+
+#### Example of a Cascading Delegation
+
+Suppose a situation where:
+
+1) End user Foobar creates a deployment.
+2) Deployment controller creates a replicaset in response.
+3) ReplicaSet controller creates a pod in response.
+
+In a delegated authority model, the delegation graph looks like:
+
+      Foobar -> Deployment controller -> ReplicaSet controller
+
+Each edge of this graph represents a single authority delegation. Controllers
+use the authority of the delegation represented by the incoming edge.
+
+#### Semantics of selecting a DelegatedAuthority
+
+Delegations encodes a permission delegation at a point in time. Revoking
+permission of the end user does not revoke permission of historic delegations
+made by the end user. If team member Foobar leaves the team, all the production
+deployments Foobar created while on the team should continue to function.
+However, actions taken using authority that was historically delegated by Foobar
+are linked to Foobar in audit via the delegation history. Similarly, persistent
+authority delegations made by Foobar while they were on the team are queryable
+by querying the list of authority delegations.
+
+Right now, users can create replica sets without necessarily being able to
+create pods. In that case, the end user lacks the authority that the replicaset
+controller would need to operate.
```

Comment:
So is the proposal that users should be granted the create-pod permission? I'm not sure that's desirable. For instance, maybe my org has a policy that users can't create pods (because that's a best-practice), and are just allowed to create deployments.